fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,979 @@
+
+        # Also check that the type of adversary_model is the same as predictor_model.
+        if isinstance(predictor_model, list):
+            # For now, if predictor_model is a list, so must adversary_model be.
+            if not isinstance(adversary_model, list):
+                raise ValueError(_KWARG_ERROR_MESSAGE.format(
+                    "adversary_model", "a list"))
+            self._initialized_models = False
+        elif self.torch:
+            if not isinstance(adversary_model, torch.nn.Module):
+                raise ValueError(_KWARG_ERROR_MESSAGE.format(
+                    "adversary_model", "a \'torch.nn.Module\'"))
+            self._initialized_models = True
+        elif self.tensorflow:
+            if not isinstance(adversary_model, tf.keras.Model):
+                raise ValueError(_KWARG_ERROR_MESSAGE.format(
+                    "adversary_model", "a \'tensorflow.keras.Model\'"))
+            self._initialized_models = True
+        # Note, if initialized_models == False, the models will still be lists
+
+        self.predictor_model = predictor_model
+        self.adversary_model = adversary_model
+
+    def _verify_dist_type(self, input_type, dist_name):
+        if callable(input_type):
+            return input_type
+        elif isinstance(input_type, str):
+            if input_type.lower() in Keyword.TYPES:
+                return input_type.lower()
+        raise ValueError(_KWARG_ERROR_MESSAGE.format(
+            dist_name, "one of {} or a callable".format(Keyword.TYPES)))
+
+    def _init_losses(self, predictor_loss, adversary_loss, predictor_function):
+        self.predictor_loss = self._verify_dist_type(predictor_loss, "predictor_loss")
+        self.adversary_loss = self._verify_dist_type(adversary_loss, "adversary_loss")
+        self.predictor_function = self._verify_dist_type(predictor_function, "predictor_function")
+
+    def _init_constraints(self, constraints):
+        """Verify the constraints and set up the corresponding network structure."""
+        self.constraints = constraints
+        if (constraints == "demographic_parity"):
+            self.pass_y = False
+        elif (constraints == "equalized_odds"):
+            self.pass_y = True
+        else:
+            raise ValueError(_KWARG_ERROR_MESSAGE.format(
+                "constraints", "\'demographic_parity\' or \'equalized_odds\'"))
+
+    def _init_cuda(self, cuda):
+        """Verify whether we can use the GPU and move pytorch model to it."""
+        if (not cuda):
+            self.cuda = False
+        elif (cuda):
+            if (not self.torch):
+                raise ValueError("Cuda is PyTorch-only")
+            if not torch.cuda.is_available():
+                raise ValueError("Cuda is not available")
+            self.cuda = True
+            self.device = torch.device("cuda:0")
+
+    def _extend_instance(self, cls_):
```

Comment:
instead of this, we can have a `BackendEngine` kinda thing, which implements the methods, and we instantiate one in `fit`, and use its methods, instead of adding them here to `self`. I think it'd be much cleaner.